drivers/SmartThings/zigbee-switch/src/inovelli-vzm32-sn/init.lua

```diff
@@ -0,0 +1,520 @@
+      return true, subdriver
+    end
+  end
+  return false
+end
+
+local function to_boolean(value)
+  if value == 0 or value =="0" then
+    return false
+  else
+    return true
+  end
+end
+
+local map_key_attribute_to_capability = {
+  [0x00] = capabilities.button.button.pushed,
+  [0x01] = capabilities.button.button.held,
+  [0x02] = capabilities.button.button.down_hold,
+  [0x03] = capabilities.button.button.pushed_2x,
+  [0x04] = capabilities.button.button.pushed_3x,
+  [0x05] = capabilities.button.button.pushed_4x,
+  [0x06] = capabilities.button.button.pushed_5x,
+}
+
+local key_mmwave_preferences = {
+  "parameter101",
+  "parameter102",
+  "parameter103",
+  "parameter104",
+  "parameter105",
+  "parameter106",
+  "parameter111",
+  "parameter112",
+  "parameter113",
+  "parameter114",
+}
+
+local function button_to_component(buttonId)
+  if buttonId > 0 then
+    return string.format("button%d", buttonId)
+  end
+end
+
+local function scene_handler(driver, device, zb_rx)
+  local bytes = zb_rx.body.zcl_body.body_bytes
+  local button_number = bytes:byte(1)
+  local capability_attribute = map_key_attribute_to_capability[bytes:byte(2)]
+  local additional_fields = {
+    state_change = true
+  }
+
+  local event
+  if capability_attribute ~= nil then
+    event = capability_attribute(additional_fields)
+  end
+
+  local comp = device.profile.components[button_to_component(button_number)]
+  if comp ~= nil then
+    device:emit_component_event(comp, event)
+  end
+end
+
+local function add_child(driver,parent,profile,child_type)
+  local child_metadata = {
+      type = "EDGE_CHILD",
+      label = string.format("%s %s", parent.label, child_type:gsub("(%l)(%w*)", function(a,b) return string.upper(a)..b end)),
+      profile = profile,
+      parent_device_id = parent.id,
+      parent_assigned_child_key = child_type,
+      vendor_provided_label = string.format("%s %s", parent.label, child_type:gsub("(%l)(%w*)", function(a,b) return string.upper(a)..b end))
+  }
+  driver:try_create_device(child_metadata)
+end
+
+local function contains(array, value)
+  for _, element in ipairs(array) do
+    if element == value then
+      return true
+    end
+  end
+  return false
+end
+
+local function configure_illuminance_reporting(device)
+  local min_lux_change = 15
+  local value = math.floor(10000 * math.log10(min_lux_change) + 1)
+  device:send(clusters.IlluminanceMeasurement.attributes.MeasuredValue:configure_reporting(
+      device,
+      10,   -- Minimum reporting interval (seconds)
+      600,  -- Maximum reporting interval (seconds)
+      value    -- Reportable change (in raw unit values)
+  ))
+end
+
+local function send_ota_image_notify(device)
+  local PAYLOAD_TYPE = 0x00
+  local QUERY_JITTER = 100
+  local MFG_CODE     = MFG_CODE
+  local IMAGE_TYPE   = 0xFFFF
+  local NEW_VERSION  = 0xFFFFFFFF
+  device:send(OTAUpgrade.commands.ImageNotify(device, PAYLOAD_TYPE, QUERY_JITTER, MFG_CODE, IMAGE_TYPE, NEW_VERSION))
+end
```

Comment:
this should be handled at a lower level by the hub automatically

Reply:
This was added because the device does not send a query next image during inclusion, but sends it every 1 hour afterwards. So in the certification test suite, it was failing the "Firmware Update" test because the device didn't have firmware data saved from a query next image. This was used so that SmartThings had the image information saved so it would pass the Firmware Update in the test suite. It can be removed, but might not pass that part of the test anymore.

Reply:
Understood. We've recently made changes to do this more aggressively on device join, but those changes are not fully rolled out yet. You should leave it in if you need it to pass cert.